libcasr/src/csharp.rs

```diff
@@ -0,0 +1,128 @@
+//! C Sharp module implements `ParseStacktrace` and `Exception` traits for C Sharp reports.
+use crate::exception::Exception;
+use crate::stacktrace::{ParseStacktrace, Stacktrace};
+
+use crate::error::*;
+use crate::execution_class::ExecutionClass;
+use crate::stacktrace::StacktraceEntry;
+
+use regex::Regex;
+
+const PARSE_STACK_FRAME_ERR_PREFIX: &str = "Couldn't parse stacktrace line: ";
+
+/// Structure provides an interface for processing the stack trace.
+pub struct CSharpStacktrace;
+
+impl CSharpStacktrace {
+    fn parse_stacktrace_entry(entry: &str, use_mono_format: bool) -> Result<StacktraceEntry> {
+        let re =
+            if use_mono_format { Regex::new(r"^at (?:\(.+\) )?(.+?) ?\(.*\) (?:<0x(?<address>\w+).+>|\[\w+]) in (?:<\w+>|(?<file>.+)):(?<line>\w+)$") }
+            else { Regex::new(r"^at (?:\(.+\) )?(.+?)\(.*\)(?: in (?<file>.+):line (?<line>\w+))?$") }.unwrap();
+
+        let Some(cap) = re.captures(entry) else {
+            return Err(Error::Casr(format!("{PARSE_STACK_FRAME_ERR_PREFIX}{entry}")));
+        };
+
+        let mut stentry = StacktraceEntry::default();
+
+        if let Some(file) = cap.name("file").map(|m| m.as_str().to_string()) {
+            let Some(line) = cap.name("line").and_then(|m| m.as_str().parse::<u64>().ok()) else {
+                return Err(Error::Casr(format!("Couldn't parse stacktrace line num: {entry}")));
+            };
+            stentry.debug.file = file;
+            stentry.debug.line = line;
+        }
+
+        if let Some(num) = cap.name("address").map(|m| m.as_str()) {
+            if let Ok(address) = u64::from_str_radix(num, 16) {
+                stentry.address = address;
+            } else {
+                return Err(Error::Casr(format!("Couldn't parse address: {num}")));
+            }
+        }
+
+        stentry.function = cap.get(1).unwrap().as_str().to_string();
+
+        Ok(stentry)
+    }
+}
+
+impl ParseStacktrace for CSharpStacktrace {
+    fn extract_stacktrace(stream: &str) -> Result<Vec<String>> {
+        let re = Regex::new(r"(?m)^Unhandled ([Ee])xception(:\n|\. )(?:.|\n)+?((?:\n[^\S\n]*(?:at.+|--- End of inner exception stack trace ---))+)$").unwrap();
+
+        let Some(cap) = re.captures(stream).and_then(|cap|
+            ((cap.get(1).unwrap().as_str() == "E") == (cap.get(2).unwrap().as_str() == ":\n")).then_some(cap)
+        ) else {
+            return Err(Error::Casr("Couldn't find stacktrace".to_string()));
```

Comment:
Add more informative error message: mention that stack trace format not recognized.